### PR TITLE
Proj6

### PR DESCRIPTION
**Pull request by Komzpa**
Pull request to check Proj master Travis.

### CODE REVIEW

**Comment by Komzpa** [comment]
@pramsey 
```
zeros -I/usr/local/pgsql/include/server -I/usr/local/include  -fPIC -DPIC  -c -o lwgeom_pg.o lwgeom_pg.c
lwgeom_transform.c:585:1: error: 'pgstrs_get_entry' defined but not used [-Werror=unused-function]
 pgstrs_get_entry(const PjStrs *strs, int n)
 ^~~~~~~~~~~~~~~~
lwgeom_transform.c:147:1: error: 'PJErrStr' defined but not used [-Werror=unused-function]
 PJErrStr()
 ^~~~~~~~
cc1: error: unrecognized command line option '-Wno-unknown-warning-option' [-Werror]
cc1: all warnings being treated as errors
make[1]: *** [Makefile:61: lwgeom_transform.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [GNUmakefile:20: all] Error 1
make[1]: Leaving directory '/src/postgis/libpgcommon'
[logbt] saw 'make' exit with code:2 (INT)
```

**Comment by Algunenano** [comment]
If you are adding support for the new API, `ACCEPT_USE_OF_DEPRECATED_PROJ_API_H` definitions should go away and instead rely on the new headers for PROJ6+:
```
configure.ac
873:#define ACCEPT_USE_OF_DEPRECATED_PROJ_API_H 1
882:dnl For PROJ 6 add the ACCEPT_USE_OF_DEPRECATED_PROJ_API_H definition
884:    AC_DEFINE([ACCEPT_USE_OF_DEPRECATED_PROJ_API_H], [1], [Define to use the old PROJ API])

macros/ac_proj4_version.m4
22:             #define ACCEPT_USE_OF_DEPRECATED_PROJ_API_H 1

postgis_config.h.in
163:#undef ACCEPT_USE_OF_DEPRECATED_PROJ_API_H

```

File: liblwgeom/liblwgeom.h.in [hunk 1]
```diff
@@ -34,14 +34,24 @@
 #include <stdint.h>
 
 #include "../postgis_config.h"
-#include "proj_api.h"
 
-#if POSTGIS_PROJ_VERSION >= 49
-/* Enable new geodesic functions */
-#define PROJ_GEODESIC 1
+#if POSTGIS_PROJ_VERSION < 60
```
Comment (Algunenano) [inline liblwgeom/liblwgeom.h.in]: You are changing the style of this file from `tabs` to 4 spaces :oncoming_police_car: My bet is `.h.in` isn't detected by your editor as a C header.

File: liblwgeom/liblwgeom_internal.h [hunk 2]
```diff
@@ -488,5 +489,6 @@ int ptarray_npoints_in_rect(const POINTARRAY *pa, const GBOX *gbox);
 int gbox_contains_point2d(const GBOX *g, const POINT2D *p);
 int lwpoly_contains_point(const LWPOLY *poly, const POINT2D *pt);
 POINT4D* lwmpoint_extract_points_4d(const LWMPOINT* g, uint32_t* npoints, int* input_empty);
+char* lwstrdup(const char* a);
```
Comment (Algunenano) [inline liblwgeom/liblwgeom_internal.h]: Nice addition, but since you added it, can you also change the existing call to strdup (which is a bug)?
`liblwgeom/g_box.c:397:          return strdup("NULL POINTER");`

File: liblwgeom/lwgeom_transform.c [hunk 3]
```diff
@@ -44,33 +44,67 @@ to_dec(POINT4D *pt)
...
+	char *pj_errstr;
+	int rv;
+	PJ pj;
+
+	pj.pj_from = lwproj_from_string(instr);
+	if (!pj.pj_from)
+	{
+		pj_errstr = pj_strerrno(*pj_get_errno_ref());
+		if (!pj_errstr) pj_errstr = "";
+		lwerror("could not parse proj string '%s'", instr);
```
Comment (Algunenano) [inline liblwgeom/lwgeom_transform.c]: You set `pj_errstr`, which contains the proj error, and then print only the projection string. Could we have both, or at least the error as show by proj?

Comment (Algunenano) [reply liblwgeom/lwgeom_transform.c]: Same in the next conditional.

File: liblwgeom/lwgeom_transform.c [hunk 4]
```diff
@@ -44,33 +44,67 @@ to_dec(POINT4D *pt)
...
+		return LW_FAILURE;
+	}
+
+	pj.pj_to = lwproj_from_string(outstr);
+	if (!pj.pj_to)
+	{
+		pj_errstr = pj_strerrno(*pj_get_errno_ref());
+		if (!pj_errstr) pj_errstr = "";
+		lwerror("could not parse proj string '%s'", instr);
+		return LW_FAILURE;
```
Comment (Algunenano) [inline liblwgeom/lwgeom_transform.c]: If this conditional fails, `pj.pj_to` will be leaked.

**Comment by Algunenano** [comment]
Isn't pj_init_plus() also deprecated with the OLD API? It's used in https://github.com/postgis/postgis/blob/e21bcac8d7a4bcda3e5a686fab01f8d5a5a67992/liblwgeom/lwgeom_transform.c#L206

According to https://github.com/OSGeo/proj.4/blob/master/docs/source/development/migration.rst#L170 it should be `proj_create` now.

File: liblwgeom/liblwgeom_internal.h (hunk 2 above)
Comment (Algunenano) [reply liblwgeom/liblwgeom_internal.h]: Nevermind, I see that you've already done it.

File: liblwgeom/lwgeom_transform.c [hunk 5]
```diff
@@ -172,3 +205,238 @@ lwproj_from_string(const char *str1)
...
+	POINT4D p;
+	size_t n_converted;
+	size_t n_points = pa->npoints;
+	size_t point_size = ptarray_point_size(pa);
+	int has_z = ptarray_has_z(pa);
+	double *pa_double = (double*)(pa->serialized_pointlist);
+
+	/* XXXX TODO check that the PJ has decimal degrees as units in input/output */
+
+	int input_swapped = proj_crs_is_swapped(proj_get_source_crs(NULL, pj));
```
Comment (Algunenano) [inline liblwgeom/lwgeom_transform.c]: I'm getting a coredump here while running the tests. `proj_get_source_crs` can return NULL and should be checked before calling `proj_get_source_crs`:
```
(gdb) bt
#0  0x00007f2beb28dd7f in raise () from /usr/lib/libc.so.6
#1  0x00007f2beb278672 in abort () from /usr/lib/libc.so.6
#2  0x00007f2beb278548 in __assert_fail_base.cold.0 () from /usr/lib/libc.so.6
#3  0x00007f2beb286396 in __assert_fail () from /usr/lib/libc.so.6
#4  0x00007f29da805062 in proj_crs_get_coordinate_system (ctx=0x7f29da8fa7c0 <pj_get_default_ctx::default_context>, crs=<optimized out>) at /usr/include/c++/8.2.1/ext/atomicity.h:69
#5  0x00007f29daa224e8 in proj_crs_is_swapped (pj_crs=0x0) at lwgeom_transform.c:294
#6  ptarray_transform (pa=0x55cf62e4b4f0, pj=0x55cf6365e7f0) at lwgeom_transform.c:330
#7  0x00007f29daa22359 in lwgeom_transform (geom=0x55cf62e4b538, pj=0x55cf6365e7f0) at lwgeom_transform.c:248
#8  0x00007f29da9754f7 in transform (fcinfo=<optimized out>) at lwgeom_transform.c:83
#9  0x000055cf627be130 in fmgr_security_definer (fcinfo=0x55cf62e495d0) at fmgr.c:748
#10 0x000055cf624e93e0 in ExecInterpExpr (state=<optimized out>, econtext=<optimized out>, isnull=0x7ffda3a9329f) at execExprInterp.c:678
#11 0x000055cf625b55ae in ExecEvalExprSwitchContext (state=<optimized out>, econtext=0x7ffda3a92bb0, isNull=0x87e417277d840000) at ../../../../src/include/executor/executor.h:303
#12 evaluate_expr (expr=<optimized out>, result_type=3353058, result_typmod=-1, result_collation=0) at clauses.c:4970
#13 0x000055cf625b4ddf in evaluate_function (funcid=3353401, result_type=<optimized out>, result_typmod=-1, result_collid=0, input_collid=0, args=0x55cf62f84900, context=<optimized out>, funcvariadic=<optimized out>, func_tuple=<optimized out>) at clauses.c:4512
#14 simplify_function (funcid=3353401, result_type=<optimized out>, result_typmod=-1, result_collid=0, input_collid=0, args_p=<optimized out>, funcvariadic=<optimized out>, process_args=<optimized out>, allow_non_const=true, context=0x7ffda3a937e0) at clauses.c:4152
#15 0x000055cf625b2fd1 in eval_const_expressions_mutator (node=<optimized out>, context=<optimized out>) at clauses.c:2742
```

Comment (Algunenano) [reply liblwgeom/lwgeom_transform.c]: Also, according to the documentation of `proj_get_source_crs`:
```
 * The returned object must be unreferenced with proj_destroy() after
 * use.
```

So you are leaking memory here.

File: liblwgeom/lwgeom_transform.c [hunk 6]
```diff
@@ -172,3 +205,238 @@ lwproj_from_string(const char *str1)
...
+{
+	PJ *pj = proj_create_crs_to_crs(NULL, instr, outstr, NULL);
+	if (!pj)
+	{
+		PJ *pj_in = proj_create(NULL, instr);
+		PJ *pj_out = proj_create(NULL, outstr);
+		if (!pj_in)
+			lwerror("could not parse proj string '%s'", instr);
+		if (!pj_out)
+			lwerror("could not parse proj string '%s'", outstr);
```
Comment (Algunenano) [inline liblwgeom/lwgeom_transform.c]: This leaks memory if pj_in succeeds and pj_out fails.

File: liblwgeom/lwgeom_transform.c [hunk 7]
```diff
@@ -172,3 +205,238 @@ lwproj_from_string(const char *str1)
...
+	return lwgeom_transform(geom, pj);
+}
+
+int
+lwgeom_transform(LWGEOM* geom, PJ* pj)
+{
+	uint32_t i;
+
+	/* No points to transform in an empty! */
+	if (lwgeom_is_empty(geom))
```
Comment (Algunenano) [inline liblwgeom/lwgeom_transform.c]: Doesn't it make more sense to check this before calling `lwgeom_transform` to avoid all the work involved in `GetPJUsingFCInfo`?

File: liblwgeom/lwgeom_transform.c [hunk 8]
```diff
@@ -172,3 +205,238 @@ lwproj_from_string(const char *str1)
...
+		{
+			getPoint4d_p(pa, i, &p);
+			to_rad(&p);
+		}
+	}
+
+	if (input_swapped)
+		ptarray_swap_ordinates(pa, LWORD_X, LWORD_Y);
+
+	// size_t proj_trans_generic(PJ *P, PJ_DIRECTION direction,
```
Comment (Algunenano) [inline liblwgeom/lwgeom_transform.c]: I was going to complain about `//`, but it appears we are using them in other places too :+1:

File: liblwgeom/lwgeom_transform.c [hunk 9]
```diff
@@ -172,3 +205,238 @@ lwproj_from_string(const char *str1)
...
+	{
+		lwerror("ptarray_transform: converted (%d) != input (%d)",
+			n_converted, n_points);
+		return LW_FAILURE;
+	}
+
+	int pj_errno_val = proj_errno(pj);
+	if (pj_errno_val)
+	{
+		lwerror("transform: %s (%d)",
```
Comment (Algunenano) [inline liblwgeom/lwgeom_transform.c]: Should `transform` be `ptarray_transform` as in the error above? In any case, both could be changed by `%s` + `__func__`

File: liblwgeom/lwgeom_transform.c [hunk 10]
```diff
@@ -172,3 +205,238 @@ lwproj_from_string(const char *str1)
...
+			getPoint4d_p(pa, i, &p);
+			to_dec(&p);
+		}
+	}
+
+	return LW_SUCCESS;
+}
+
+int
+point4d_transform(POINT4D *pt, PJ* pj)
```
Comment (Algunenano) [inline liblwgeom/lwgeom_transform.c]: I don't see this function being used anywhere except for the old `ptarray_transform`. In that case, I suggest removing it from liblwgeom_internal.h, making it static and removing the copy for `/* POSTGIS_PROJ_VERION >= 60 */`

File: libpgcommon/lwgeom_transform.c [hunk 11]
```diff
@@ -220,19 +276,19 @@ static HTAB *CreatePJHash(void)
...
+	return hash_create("PostGIS PROJ Backend MemoryContext Hash", PROJ_BACKEND_HASH_SIZE, &ctl, (HASH_ELEM | HASH_FUNCTION));
 }
 
-static void AddPJHashEntry(MemoryContext mcxt, projPJ projection)
+static void AddPJHashEntry(MemoryContext mcxt, PJ* projection)
 {
 	bool found;
-	void **key;
-	PJHashEntry *he;
+	void** key;
```
Comment (Algunenano) [inline libpgcommon/lwgeom_transform.c]: Why did you change the position of the `*` in these functions? AFAIK (and clang-format applies) it's `type *name`

File: libpgcommon/lwgeom_transform.c [hunk 12]
```diff
@@ -451,102 +521,164 @@ static char* GetProj4String(int srid)
...
+	if (!projection->pj_to)
+		elog(ERROR,
+		    "AddToPROJSRSCache: could not form projection from 'srid=%d' to 'srid=%d'",
+		    srid_from, srid_to);
+#else
+	PJ* projection = NULL;
+	/* Try combinations of ESPG/SRTEXT/PROJ4TEXT until we find */
+	/* one that gives us a usable transform. Note that we prefer */
+	/* EPSG numbers over SRTEXT and SRTEXT over PROJ4TEXT */
+	for (uint32_t i = 0; i < 9; i++)
```
Comment (Algunenano) [inline libpgcommon/lwgeom_transform.c]: It'd nice if the comment explained that 9 comes from 3^2. I know it says it more or less, but at first I thought it was `PROJ_CACHE_ITEMS` + 1-

BTW, since this is compiled with Postgres flags, expect compiler complaints about declaring variables inside the `for` for Postgres releases before 11, for example this will break with PG10+PROJ6+Werror. `been there, done that`

File: libpgcommon/lwgeom_transform.c [hunk 13]
```diff
@@ -451,102 +521,164 @@ static char* GetProj4String(int srid)
...
 	** Turn the SRID number into a proj4 string, by reading from spatial_ref_sys
 	** or instantiating a magical value from a negative srid.
 	*/
-	proj_str = GetProj4String(srid);
-	if ( ! proj_str )
+	from_strs = GetProjStrings(srid_from);
+	to_strs = GetProjStrings(srid_to);
+	if (!pjstrs_has_entry(&from_strs))
+		elog(ERROR, "AddToPROJSRSCache: GetProjStrings returned NULL for SRID (%d)", srid_from);
+	if (!pjstrs_has_entry(&to_strs))
```
Comment (Algunenano) [inline libpgcommon/lwgeom_transform.c]: This might leak it `from_strs` is successful and `to_strs` isn't

File: libpgcommon/lwgeom_transform.c [hunk 14]
```diff
@@ -666,147 +797,144 @@ void SetPROJ4LibPath(void)
...
+	double major_axis, minor_axis, eccentricity_squared;
+#endif
+
+	if ( GetPJUsingFCInfo(fcinfo, srid, srid, &pj) == LW_FAILURE)
+		return LW_FAILURE;
+
+#if POSTGIS_PROJ_VERSION >= 60
+	if (!proj_pj_is_latlong(pj))
+		return LW_FAILURE;
+	pj_crs = proj_get_source_crs(NULL, pj);
```
Comment (Algunenano) [inline libpgcommon/lwgeom_transform.c]: `proj_get_source_crs` and `proj_get_ellipsoid` can return NULL, which would crash subsequent function calls.

File: liblwgeom/liblwgeom.h.in (hunk 1 above)
Comment (pramsey) [reply liblwgeom/liblwgeom.h.in]: Editorconfig has `h.in` set to spaces... not saying this is right, just saying it's what the file seems to be

File: liblwgeom/lwgeom_transform.c (hunk 3 above)
Comment (pramsey) [reply liblwgeom/lwgeom_transform.c]: Yeah... no. The issue here is the classic: upstream library has changed their error strings. I found I couldn't pass regression when I was writing out the full error strings. Quick judgement call: strip regression tests or add string?

**Comment by pramsey** [comment]
`pj_init_plus` is part of the proj4 API and it's call in the < 6 code block, so I feel "OK" about it remaining. I'm treating proj5 as == to proj4, since it happily accepts calls against the old `proj_api.h`

File: liblwgeom/lwgeom_transform.c (hunk 3 above)
Comment (Algunenano) [reply liblwgeom/lwgeom_transform.c]: 3rd option: rewrite the test so it matches the error with a regex. I think I've done this somewhere in regress/ (but i don't remember where): auxiliar function that receives 2 strings (a query and a regex), executes the query and returns true if an exception is thrown and it matches the error, false otherwise.

File: liblwgeom/liblwgeom.h.in (hunk 1 above)
Comment (Algunenano) [reply liblwgeom/liblwgeom.h.in]: I'd never checked the contents of editorconfig. I'm not sure if I would dare to do it again.

File: liblwgeom/lwgeom_transform.c (hunk 3 above)
Comment (pramsey) [reply liblwgeom/lwgeom_transform.c]: Please show me where... if it where a return value we could use a postgres regex on it, but since it's an ERROR string, I think we are SOL.

File: liblwgeom/lwgeom_transform.c (hunk 5 above)
Comment (pramsey) [reply liblwgeom/lwgeom_transform.c]: I wonder what you're running in tests that gets you a NULL? Will see if I get an error in regress now that I'm guarding against that.

File: liblwgeom/lwgeom_transform.c (hunk 3 above)
Comment (Algunenano) [reply liblwgeom/lwgeom_transform.c]: It doesn't do any regex, but it might do the trick here: https://github.com/postgis/postgis/blob/515f55bafe0d10d21c199091e3bc500491a72b31/topology/test/regress/st_remedgemodface.sql#L140

File: liblwgeom/lwgeom_transform.c (hunk 7 above)
Comment (pramsey) [reply liblwgeom/lwgeom_transform.c]: Mmm, I like the cleanliness of the check here...

File: liblwgeom/lwgeom_transform.c (hunk 5 above)
Comment (Algunenano) [reply liblwgeom/lwgeom_transform.c]: It happened when running regress/core/ but I don't recall the exact test now. I can retry in a different PC tomorrow if you want (same OS and libraries, but different PC) or in that same machine on Friday if you need.

File: liblwgeom/lwgeom_transform.c (hunk 9 above)
Comment (pramsey) [reply liblwgeom/lwgeom_transform.c]: again I'm trying to match up messages across different proj backends... I can just remove the prefix entirely...

File: liblwgeom/lwgeom_transform.c (hunk 10 above)
Comment (pramsey) [reply liblwgeom/lwgeom_transform.c]: I'll  #ifdef it out of 6+ and static-ize it as you note. I don't want to fully delete it, as it has some maybe future-useful code in it.

File: libpgcommon/lwgeom_transform.c (hunk 11 above)
Comment (pramsey) [reply libpgcommon/lwgeom_transform.c]: I'm just OCD'ing on that lately, since pushed GEOS to type-bound in the big reformat.

File: libpgcommon/lwgeom_transform.c [hunk 15]
```diff
@@ -666,147 +825,153 @@ void SetPROJ4LibPath(void)
...
+static int
+proj_pj_is_latlong(const PJ* pj)
+{
+#if POSTGIS_PROJ_VERSION < 60
+	return pj_is_latlong(pj->pj_from);
 #else
-	/* For old versions of Proj we cannot lookup the spheroid parameters from the API */
-	/* So we use the WGS84 parameters (boo!) */
-	spheroid_init(s, WGS84_MAJOR_AXIS, WGS84_MINOR_AXIS);
+	PJ_TYPE pj_type = proj_get_type(proj_get_source_crs(NULL, pj));
```
Comment (Algunenano) [inline libpgcommon/lwgeom_transform.c]: This will crash if `proj_get_source_crs` returns NULL:

When running regress/core/geography:
```
#0  0x00007ff96c529d7f in raise () from /usr/lib/libc.so.6
#1  0x00007ff96c514672 in abort () from /usr/lib/libc.so.6
#2  0x00007ff96c514548 in __assert_fail_base.cold.0 () from /usr/lib/libc.so.6
#3  0x00007ff96c522396 in __assert_fail () from /usr/lib/libc.so.6
#4  0x00007ff75bb8fb05 in proj_get_type (obj=<optimized out>) at iso19111/c_api.cpp:904
#5  0x00007ff75bd5764c in proj_pj_is_latlong (pj=0x557002aacea0) at lwgeom_transform.c:874
#6  spheroid_init_from_srid (fcinfo=<optimized out>, srid=<optimized out>, s=0x7fffa99d5828) at lwgeom_transform.c:939
#7  0x00007ff75bd1f8c8 in geography_distance (fcinfo=0x5570027d3ec0) at geography_measurement.c:223
#8  0x00005570003243e0 in ExecInterpExpr (state=<optimized out>, econtext=<optimized out>, isnull=0x7fffa99d599f) at execExprInterp.c:678
#9  0x0000557000356bf9 in ExecEvalExprSwitchContext (state=0x5570027d3de8, econtext=0x5570027d3b10, isNull=0x3242a5f189c500da)
    at ../../../src/include/executor/executor.h:303
#10 ExecProject (projInfo=0x5570027d3de0) at ../../../src/include/executor/executor.h:337
#11 ExecResult (pstate=<optimized out>) at nodeResult.c:136
#12 0x000055700032c200 in ExecProcNode (node=<optimized out>) at ../../../src/include/executor/executor.h:237
#13 ExecutePlan (estate=0x5570027d37f0, planstate=0x5570027d3a00, use_parallel_mode=<optimized out>, operation=CMD_SELECT, 
    sendTuples=<optimized out>, numberTuples=1, direction=ForwardScanDirection, dest=0x5570027d2510, execute_once=<optimized out>)
    at execMain.c:1723
```

Comment (Algunenano) [reply libpgcommon/lwgeom_transform.c]: It also leaks memory:
```
 * @return Object that must be unreferenced with proj_destroy(), or NULL
 * in case of error, or missing source CRS.
 */
PJ *proj_get_source_crs(PJ_CONTEXT *ctx, const PJ *obj) {
```

File: liblwgeom/lwgeom_transform.c [hunk 16]
```diff
@@ -126,49 +198,268 @@ lwgeom_transform(LWGEOM *geom, projPJ inpj, projPJ outpj)
...
+	int has_z = ptarray_has_z(pa);
+	double *pa_double = (double*)(pa->serialized_pointlist);
+	int input_swapped, output_swapped;
+
+	/* XXXX TODO check that the PJ has decimal degrees as units in input/output */
+
+	PJ* pj_source_crs = proj_get_source_crs(NULL, pj);
+	PJ* pj_target_crs = proj_get_target_crs(NULL, pj);
+
+	if (!(pj_source_crs && pj_source_crs))
```
Comment (Algunenano) [inline liblwgeom/lwgeom_transform.c]: This is checking the same pointer twice. Should it be `pj_source_crs && pj_target_crs`?
In any case, if only one of them failed the other needs to be destroyed to avoid leaks.

File: liblwgeom/lwgeom_transform.c [hunk 17]
```diff
@@ -126,49 +198,268 @@ lwgeom_transform(LWGEOM *geom, projPJ inpj, projPJ outpj)
...
+	}
+
 	return LW_SUCCESS;
 }
 
-projPJ
-lwproj_from_string(const char *str1)
+#if 0
+int
+point4d_transform(POINT4D *pt, PJ* pj)
```
Comment (Algunenano) [inline liblwgeom/lwgeom_transform.c]: Since you want to keep it, you need to handle errors in `proj_get_source_crs` as it might return NULL and change the behaviour of proj_crs_is_swapped (the default context will be used, instead of what you'd expect).

It also leaks memory.

File: liblwgeom/lwgeom_transform.c (hunk 5 above)
Comment (Algunenano) [reply liblwgeom/lwgeom_transform.c]: This happened when running `../loader/ReprojectPts`, now it doesn't crash and simply fails. I think this is an issue with my environment (I wasn't packing the datumgrids in my custom packages), but test failures are ok, crashing wasn't.

Comment (Algunenano) [reply liblwgeom/lwgeom_transform.c]: I can confirm that once I've rebuilt with current master and adding the datumgrids, the tests pass.

**Comment by Komzpa** [comment]
Travis masters failed:
https://travis-ci.org/postgis/postgis/jobs/496251295#L1123
```


gserialized_gist_2d.c: In function 'box2df_is_empty':
gserialized_gist_2d.c:147:6: error: implicit declaration of function 'isnan' [-Werror=implicit-function-declaration]
  if (isnan(a->xmin))
      ^~~~~
gserialized_gist_2d.c: In function 'box2df_set_empty':
gserialized_gist_2d.c:155:42: error: 'NAN' undeclared (first use in this function)
  a->xmin = a->xmax = a->ymin = a->ymax = NAN;
                                          ^~~
gserialized_gist_2d.c:155:42: note: each undeclared identifier is reported only once for each function it appears in
gserialized_gist_2d.c: In function 'box2df_set_finite':
gserialized_gist_2d.c:161:9: error: implicit declaration of function 'isfinite'; did you mean 'isdigit'? [-Werror=implicit-function-declaration]
  if ( ! isfinite(a->xmax) )
         ^~~~~~~~
         isdigit
gserialized_gist_2d.c: In function 'box2df_distance_leaf_centroid':
gserialized_gist_2d.c:489:12: error: implicit declaration of function 'sqrt' [-Werror=implicit-function-declaration]
     return sqrt((a_x - b_x) * (a_x - b_x) + (a_y - b_y) * (a_y - b_y));
            ^~~~
gserialized_gist_2d.c:489:12: error: incompatible implicit declaration of built-in function 'sqrt' [-Werror]
gserialized_gist_2d.c:489:12: note: include '<math.h>' or provide a declaration of 'sqrt'
gserialized_gist_2d.c:52:1:
+#include <math.h>
 
gserialized_gist_2d.c:489:12:
     return sqrt((a_x - b_x) * (a_x - b_x) + (a_y - b_y) * (a_y - b_y));
            ^~~~
gserialized_gist_2d.c: In function 'pt_distance':
gserialized_gist_2d.c:495:9: error: incompatible implicit declaration of built-in function 'sqrt' [-Werror]
  return sqrt((ax - bx) * (ax - bx) + (ay - by) * (ay - by));
         ^~~~
gserialized_gist_2d.c:495:9: note: include '<math.h>' or provide a declaration of 'sqrt'
gserialized_gist_nd.c: In function 'gidx_distance':
gserialized_gist_nd.c:604:8: error: implicit declaration of function 'isfinite'; did you mean 'isdigit'? [-Werror=implicit-function-declaration]
   if (!isfinite(d))
        ^~~~~~~~
        isdigit
gserialized_gist_nd.c:612:9: error: implicit declaration of function 'sqrt' [-Werror=implicit-function-declaration]
  return sqrt(sum);
         ^~~~
gserialized_gist_nd.c:612:9: error: incompatible implicit declaration of built-in function 'sqrt' [-Werror]
gserialized_gist_nd.c:612:9: note: include '<math.h>' or provide a declaration of 'sqrt'
gserialized_gist_nd.c:52:1:
+#include <math.h>
 
gserialized_gist_nd.c:612:9:
  return sqrt(sum);
         ^~~~
gserialized_gist_nd.c: In function 'gserialized_distance_nd':
gserialized_gist_nd.c:754:14: error: incompatible implicit declaration of built-in function 'sqrt' [-Werror]
   distance = sqrt(distance);
              ^~~~
gserialized_gist_nd.c:754:14: note: include '<math.h>' or provide a declaration of 'sqrt'
gserialized_gist_2d.c: In function 'gserialized_gist_picksplit_2d':
gserialized_gist_2d.c:2044:13: error: implicit declaration of function 'ceil' [-Werror=implicit-function-declaration]
   int   m = ceil(LIMIT_RATIO * (double) nentries);
             ^~~~
gserialized_gist_2d.c:2044:13: error: incompatible implicit declaration of built-in function 'ceil' [-Werror]
gserialized_gist_2d.c:2044:13: note: include '<math.h>' or provide a declaration of 'ceil'
In file included from ../libpgcommon/lwgeom_pg.h:19,
                 from gserialized_gist_nd.c:47:
gserialized_gist_nd.c: In function 'gserialized_gist_picksplit':
echo "BEGIN;" > postgis_upgrade.sql
/usr/local/pgsql/include/server/utils/geo_decls.h:34:23: error: implicit declaration of function 'fabs' [-Werror=implicit-function-declaration]
 #define FPeq(A,B)    (fabs((A) - (B)) <= EPSILON)
                       ^~~~
gserialized_gist_nd.c:1761:14: note: in expansion of macro 'FPeq'
     else if (FPeq(center, avgCenter[d]))
              ^~~~
/usr/local/pgsql/include/server/utils/geo_decls.h:34:23: error: incompatible implicit declaration of built-in function 'fabs' [-Werror]
 #define FPeq(A,B)    (fabs((A) - (B)) <= EPSILON)
                       ^~~~
gserialized_gist_nd.c:1761:14: note: in expansion of macro 'FPeq'
     else if (FPeq(center, avgCenter[d]))
              ^~~~
/usr/local/pgsql/include/server/utils/geo_decls.h:34:23: note: include '<math.h>' or provide a declaration of 'fabs'
 #define FPeq(A,B)    (fabs((A) - (B)) <= EPSILON)
                       ^~~~
gserialized_gist_nd.c:1761:14: note: in expansion of macro 'FPeq'
     else if (FPeq(center, avgCenter[d]))
              ^~~~
cat postgis_drop_before.sql postgis_upgrade.sql.in postgis_drop_after.sql >> postgis_upgrade.sql
echo "COMMIT;" >> postgis_upgrade.sql
#strip BEGIN/COMMIT since these are not allowed in extensions
# WARNING: we should make sure these are not internal to function bodies
gserialized_gist_2d.c: At top level:
cat postgis_drop_before.sql postgis_upgrade_for_extension.sql.in postgis_drop_after.sql | /usr/bin/perl -pe 's/BEGIN\;//g ; s/COMMIT\;//g' > postgis_upgrade_for_extension.sql
cc1: error: unrecognized command line option '-Wno-unknown-warning-option' [-Werror]
cc1: all warnings being treated as errors
make[1]: *** [<builtin>: gserialized_gist_2d.o] Error 1
make[1]: *** Waiting for unfinished jobs....
gserialized_spgist_2d.c: In function 'gserialized_spgist_compress_2d':
gserialized_spgist_2d.c:683:8: error: implicit declaration of function 'isfinite'; did you mean 'isdigit'? [-Werror=implicit-function-declaration]
  if ((!isfinite(bbox_out->xmax) || !isfinite(bbox_out->xmin)) ||
        ^~~~~~~~
        isdigit
gserialized_spgist_2d.c: At top level:
cc1: error: unrecognized command line option '-Wno-unknown-warning-option' [-Werror]
cc1: all warnings being treated as errors
make[1]: *** [<builtin>: gserialized_spgist_2d.o] Error 1
gserialized_gist_nd.c: At top level:
cc1: error: unrecognized command line option '-Wno-unknown-warning-option' [-Werror]
cc1: all warnings being treated as errors
make[1]: *** [<builtin>: gserialized_gist_nd.o] Error 1
make[1]: Leaving directory '/src/postgis/postgis'
make: *** [GNUmakefile:20: all] Error 1
[logbt] saw 'make' exit with code:2 (INT)
The command "docker run $ci_env --name pgtest-${TRAVIS_BUILD_NUMBER} pgtest" exited with 2.

```

**Comment by Komzpa** [comment]
The issue is still there on latest masters build: https://travis-ci.org/postgis/postgis/jobs/496612463#L1125
(you can temporarily remove it from allowed failures in .travis.yml on this branch)

**Comment by pramsey** [comment]
This is now committed, my God have mercy on our souls. 25ce5fde36e94ef0aaad859cfe9f6dd70dad8999

**Comment by Algunenano** [comment]
@pramsey You've left some comments unaddressed. Can you please have a look?:
- https://github.com/postgis/postgis/pull/375/files#diff-b020abf186a680c968971a33e7808a16R874: it is a coredump and or a memory leak.
- https://github.com/postgis/postgis/pull/375/files#r258836491: A copy and paste error; should be `pj_source_crs && pj_source_crs` and measures should be taken to free memory in case of error of just one of them.
- https://github.com/postgis/postgis/pull/375/files#r258837771: The code can coredump as you aren't checking the output of the proj_get_source_crs calls (nor freeing them). As it's behing #ifdef 0, I'd be more inclined to delete it completely (code will always be available in SVN/GIT).
